File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryResponse.java
package org.wikipedia.dataclient.mwapi;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.google.gson.annotations.SerializedName;

import java.util.Collections;
import java.util.Map;

import kotlinx.serialization.Serializable;

@Serializable
public class MwQueryResponse extends MwResponse {

    @SuppressWarnings("unused") @SerializedName("batchcomplete") private boolean batchComplete;

    @SuppressWarnings("unused") @SerializedName("continue") @Nullable private Map<String, String> continuation;

    @Nullable private MwQueryResult query;

    public boolean batchComplete() {
        return batchComplete;
    }

    @NonNull public Map<String, String> continuation() {
        return continuation != null ? continuation : Collections.emptyMap();
    }

    @Nullable public MwQueryResult query() {
        return query;
    }

    public boolean success() {
        return query != null;
    }

    @VisibleForTesting protected void setQuery(@Nullable MwQueryResult query) {
        this.query = query;
    }
}

File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwQueryResponse.kt
package org.wikipedia.dataclient.mwapi

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class MwQueryResponse(
    val error: MwServiceError? = null,
    val warnings: Warnings? = null,
    val batchcomplete: Boolean = false,
    val continue_: Continue? = null,
    val query: MwQueryResult? = null
) {
    @Serializable
    data class Warnings(
        val main: Warning? = null,
        val revisions: Warning? = null,
        val query: Warning? = null
    )

    @Serializable
    data class Warning(
        val warnings: String? = null,
        @SerialName("*") val text: String? = null
    )

    @Serializable
    data class Continue(
        val sroffset: Int = 0,
        val gsroffset: Int = 0,
        val gpsoffset: Int = 0,
        val continue_: String = "",
        val grncontinue: String = "",
        val gcmcontinue: String = "",
        val rvcontinue: String = "",
        val gsrcontinue: String = ""
    )
}
